File: javascript/ais_sdk/moderation_video.js
var https = require("https");
var utils = require("./utils");
var signer = require("./signer");
var ais = require("./ais");

function video(token, url, frame_interval, category, callback) {

    var endPoint = utils.getEndPoint(ais.MODERATION_SERVICE);
    /**
     *  url：视频的URL路径
     *  frame_interval：非必选 截帧时间间隔。number
     *  categories：非必选 检测场景 array politics：是否涉及政治人物的检测。terrorism：是否包含暴恐元素的检测。porn：是否包含涉黄内容元素的检测。
     * @type {string}
     */
    var requestData = {"url": url, "frame_interval": frame_interval, "category": category};
    var requestBody = JSON.stringify(requestData);

    // 构建请求信息
    var headers = {"Content-Type": "application/json", "X-Auth-Token": token, "Content-Length": requestBody.length};
    var options = utils.getHttpRequestEntityOptions(endPoint, "POST", ais.MODERATION_VIDEO, headers);

    var request = https.request(options, function (response) {

        response.on("data", function (chunk) {

            // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
            if (response.statusCode !== 200) {
                console.log('Http status code is: ' + response.statusCode);
                console.log(chunk.toString());
                return;
            }

            // 获取job_id
            var result = JSON.parse(chunk);
            job_id = result.result.job_id;
            console.log('Process job id is :' + job_id);
            var retryTimes = 0;

            // 根据job_id的结果,获取视频审核信息
            var results = "";
            get_result(endPoint, job_id, results, token, retryTimes, callback);
        })
    });

    request.on("error", function (err) {
        console.log(err.message);
    });

    request.write(requestBody);
    request.end();
}


function get_result(endPoint, job_id, resultsearch, token, retryTimes, callback) {
    // 构建请求参数和请求信息
    var requestData = {'job_id': job_id};
    var options = utils.getHttpRequestEntityForGet(endPoint, "GET", ais.MODERATION_VIDEO, {
        "Content-Type": "application/json",
        "X-Auth-Token": token
    }, requestData);

    // 轮询请求视频审核接口，获取结果信息
    var reqsearch = https.request(options, function (res) {
        res.setEncoding('utf8');
        res.on('data', function (chunk) {
            resultsearch = JSON.parse(chunk);
            if (res.statusCode !== 200) {
                if(retryTimes < ais.RETRY_TIMES_MAX){
                    retryTimes++;
                    // 满足条件进行重试
                    setTimeout(function () {
                        get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback);
                    }, 2000);
                    console.log('The processing job request failed, retrying');
                }else {
                    console.log('Http status code is: ' + response.statusCode);
                    console.log(chunk.toString());
                }
                return;
            }

            // 如果处理失败，直接退出
            if (resultsearch !== "" && resultsearch.result.status === "failed") {
                console.log('The processing job request failed');
                callback(JSON.stringify(resultsearch));
                return;
            }

            // 任务处理成功
            if (resultsearch !== "" && resultsearch.result.status === "finish") {
                callback(JSON.stringify(resultsearch));
            }
            else {

                // 如果没有返回，等待一段时间，继续进行轮询。
                setTimeout(function () {
                    get_result(endPoint, job_id, resultsearch, token, retryTimes, callback);
                }, 2000);
            }
        });
    });

    reqsearch.on('error', function (e) {
        console.log('problem with request: ' + e.message);
    });

    reqsearch.write("");
    reqsearch.end();
}

function video_aksk(_ak, _sk, url, frame_interval, category, callback) {
    // 配置ak，sk信息
    var sig = new signer.Signer();
    sig.AppKey = _ak;                     // 构建ak
    sig.AppSecret = _sk;                  // 构建sk

    var endPoint = utils.getEndPoint(ais.MODERATION_SERVICE);

    // 构建请求信息和请求参数信息
    var requestData = {"url": url, "frame_interval": frame_interval, "category": category};
    var job_id = "";
    var req = new signer.HttpRequest();
    var header = {"Content-Type": "application/json"};
    var options = utils.getHttpRequestEntity(sig, req, endPoint, "POST", ais.MODERATION_VIDEO, "", header, requestData);

    var request = https.request(options, function (response) {

        response.on("data", function (chunk) {

            // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
            if (response.statusCode !== 200) {
                console.log('Http status code is: ' + response.statusCode);
                console.log(chunk.toString());
                return;
            }

            // 获取job_id 信息，获取视频审核信息
            var result = JSON.parse(chunk);
            job_id = result.result.job_id;
            console.log('Process job id is :' + job_id);
            var retryTimes = 0;

            var results = "";
            get_result_aksk(endPoint, sig, job_id, results, retryTimes, callback);
        })
    });

    request.on("error", function (err) {
        console.log(err.message);
    });

    request.write(req.body);
    request.end();
}


/**
 * 通过job_id 的信息，获取视频审核结果
 * @param endPoint ：服务域名
 * @param sign : 签名对象
 * @param job_id : 任务的id
 * @param resultsearch ：获取结果信息
 * @param retryTimes ：失败重试次数
 * @param callback 回调函数
 */
function get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback) {

    // 构建请求信息和参数信息
    var request = new signer.HttpRequest();
    var options = utils.getHttpRequestEntity(sign, request, endPoint, "GET", ais.MODERATION_VIDEO, {'job_id': job_id}, {"Content-Type": "application/json"}, "");

    // 轮询请求视频接口，获取结果信息
    var reqsearch = https.request(options, function (response) {

        response.setEncoding('utf8');
        response.on('data', function (chunk) {
            resultsearch = JSON.parse(chunk);
            if (response.statusCode !== 200) {
                if(retryTimes < ais.RETRY_TIMES_MAX){
                    retryTimes++;
                    // 满足条件进行重试
                    setTimeout(function () {
                        get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback);
                    }, 2000);
                    console.log('The processing job request failed, retrying');
                }else {
                    console.log('Http status code is: ' + response.statusCode);
                    console.log(chunk.toString());
                }
                return;
            }

            // 如果处理失败，直接退出
            if (resultsearch !== "" && resultsearch.result.status === "failed") {
                console.log('The processing job request failed');
                callback(JSON.stringify(resultsearch));
                return;
            }

            // 任务处理成功
            if (resultsearch !== "" && resultsearch.result.status === "finish") {
                callback(JSON.stringify(resultsearch));
            }
            else {

                // 如果没有返回，等待一段时间，继续进行轮询。
                setTimeout(function () {
                    get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback);
                }, 2000);
            }
        });
    });

    reqsearch.on('error', function (e) {
        console.log('problem with request: ' + e.message);

    });

    reqsearch.end();
}

module.exports = {
    video,
    video_aksk
};

File: python3/ais_sdk/moderation_video.py
# -*- coding:utf-8 -*-
import ssl
import ais_sdk.signer as signer
from urllib.error import URLError, HTTPError
import urllib.parse
import urllib.request
import json
import time
import ais_sdk.ais as ais
import ais_sdk.utils as utils

#
# access asr, long_sentence，post data by token
#
def moderation_video(token, url, frame_interval=5, categories=['politics', 'terrorism']):
    endpoint = utils.get_endpoint(ais.AisService.MODERATION_SERVICE)
    status, r = _moderation_video(endpoint, token, url, frame_interval, categories)

    if status != 200:
        return r

    submit_result = json.loads(r)
    job_id = submit_result['result'].get('job_id', '')
    #print("Process job id is :", job_id)
    time.sleep(1.0)
    try:
        while True:
            status, r = _get_result(endpoint, token, job_id)
            if status != 200:
                return r

            rec_result = json.loads(r)

            process_status = rec_result["result"].get('status')
            if process_status == 'failed':
                return r

            elif process_status == 'finish':
                return r

            #
            # process_status == 0 || process_status == 1
            #
            else:
                time.sleep(2.0)
                continue

    except Exception:
        return ''


#
# moderation_video, post the data
#
def _moderation_video(endpoint, token, url, frame_interval=5, categories=['politics', 'terrorism']):
    _url = 'https://%s/v1.0/moderation/video' % endpoint

    _data = {
        "url": url,
        "frame_interval": frame_interval,
        "categories": categories
    }
    _headers = {
        "Content-Type": "application/json",
        "X-Auth-Token": token
    }
    data = json.dumps(_data).encode("utf-8")

    kreq = urllib.request.Request(_url, data, _headers)

    resp = None
    status_code = None
    try:
        #
        # Here we use the unvertified-ssl-context, Because in FunctionStage
        # the client CA-validation have some problem, so we must do this.
        #
        _context = ssl._create_unverified_context()
        r = urllib.request.urlopen(kreq, context=_context)
    #
    # We use HTTPError and URLError，because urllib can't process the 4XX &
    # 500 error in the single urlopen function.
    #
    # If you use a modern, high-level designed HTTP client lib, Yeah, I mean requests,
    # there is no this problem.
    #
    except HTTPError as e:
        resp = e.read()
        status_code = e.code
    except URLError as e:
        resp = e.read()
        status_code = e.code
    else:
        status_code = r.code
        resp = r.read()
    return status_code, resp.decode('utf-8')


#
# access asr, moderation vedio, get the result
#
def _get_result(endpoint, token, job_id):
    _url_tmpl = 'https://%s/v1.0/moderation/video?job_id=%s'
    _url = _url_tmpl % (endpoint, job_id)
    _headers = {
        "Content-Type": "application/json",
        "X-Auth-Token": token
    }
    kreq = urllib.request.Request(_url, headers=_headers)

    resp = None
    status_code = None
    try:
        #
        # Here we use the unvertified-ssl-context, Because in FunctionStage
        # the client CA-validation have some problem, so we must do this.
        #
        _context = ssl._create_unverified_context()
        r = urllib.request.urlopen(kreq, context=_context)
    #
    # We use HTTPError and URLError，because urllib can't process the 4XX &
    # 500 error in the single urlopen function.
    #
    # If you use a modern, high-level designed HTTP client lib, Yeah, I mean requests,
    # there is no this problem.
    #
    except HTTPError as e:
        resp = e.read()
        status_code = e.code
    except URLError as e:
        resp = e.read()
        status_code = e.code
    else:
        status_code = r.code
        resp = r.read()
    return status_code, resp.decode('utf-8')


#
# access asr, long_sentence，post data by ak,sk
#
def moderation_video_aksk(_ak, _sk, url, frame_interval=5, categories=['politics', 'terrorism']):
    endpoint = utils.get_endpoint(ais.AisService.MODERATION_SERVICE)
    sig = signer.Signer()
    sig.AppKey = _ak
    sig.AppSecret = _sk

    status, r = _moderation_video_aksk(endpoint, sig, url, frame_interval, categories)

    if status != 200:
        return r

    submit_result = json.loads(r)
    job_id = submit_result['result'].get('job_id', '')
    #print("Process job id is :", job_id)
    time.sleep(1.0)
    try:
        while True:
            status, r = _get_result_aksk(endpoint, sig, job_id)
            if status != 200:
                return r

            rec_result = json.loads(r)

            process_status = rec_result["result"].get('status')
            if process_status == 'failed':
                return r

            elif process_status == 'finish':
                return r

            #
            # process_status == 0 || process_status == 1
            #
            else:
                time.sleep(2.0)
                continue

    except Exception:
        return ''


#
# moderation_video, post the data
#
def _moderation_video_aksk(endpoint, sig, url, frame_interval=5, categories=['politics', 'terrorism']):
    _url = 'https://%s/v1.0/moderation/video' % endpoint

    _data = {
        "url": url,
        "frame_interval": frame_interval,
        "categories": categories
    }

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = endpoint
    kreq.uri = "/v1.0/moderation/video"
    kreq.method = "POST"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.body = json.dumps(_data)

    resp = None
    status_code = None
    try:
        sig.Sign(kreq)
        #
        # Here we use the unvertified-ssl-context, Because in FunctionStage
        # the client CA-validation have some problem, so we must do this.
        #
        _context = ssl._create_unverified_context()
        req = urllib.request.Request(url=_url, data=kreq.body, headers=kreq.headers)
        r = urllib.request.urlopen(req, context=_context)
    #
    # We use HTTPError and URLError，because urllib can't process the 4XX &
    # 500 error in the single urlopen function.
    #
    # If you use a modern, high-level designed HTTP client lib, Yeah, I mean requests,
    # there is no this problem.
    #
    except HTTPError as e:
        resp = e.read()
        status_code = e.code
    except URLError as e:
        resp = e.read()
        status_code = e.code
    else:
        status_code = r.code
        resp = r.read()
    return status_code, resp.decode('utf-8')


#
# access asr, moderation vedio, get the result
#
def _get_result_aksk(endpoint, sig, job_id):
    _url_tmpl = 'https://%s/v1.0/moderation/video?job_id=%s'
    _url = _url_tmpl % (endpoint, job_id)

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = endpoint
    kreq.uri = "/v1.0/moderation/video"
    kreq.method = "GET"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.query = {'job_id': job_id}
    resp = None
    status_code = None
    try:
        sig.Sign(kreq)
        #
        # Here we use the unvertified-ssl-context, Because in FunctionStage
        # the client CA-validation have some problem, so we must do this.
        #
        _context = ssl._create_unverified_context()
        req = urllib.request.Request(url=_url, headers=kreq.headers)
        r = urllib.request.urlopen(req, context=_context)

    #
    # We use HTTPError and URLError，because urllib can't process the 4XX &
    # 500 error in the single urlopen function.
    #
    # If you use a modern, high-level designed HTTP client lib, Yeah, I mean requests,
    # there is no this problem.
    #
    except HTTPError as e:
        resp = e.read()
        status_code = e.code
    except URLError as e:
        resp = e.read()
        status_code = e.code
    else:
        status_code = r.code
        resp = r.read()
    return status_code, resp.decode('utf-8')
